### PR TITLE
Optimize filter processing in SPI thread

### PR DESCRIPTION
**Pull request by andyp1per**
This yields a fairly significant - and means the difference between poor quality flight and controlled flight - CPU improvement. On an F7, before:

```
ThreadsV2
ISR           PRI=255 sp=0x20010000 STACK=1224/1536 LOAD= 5.5%
ArduCopter    PRI=182 sp=0x20010600 STACK=5256/7168 LOAD=64.0%
idle          PRI=  1 sp=0x20023770 STACK= 144/ 352 LOAD= 0.0%
UART_RX       PRI= 60 sp=0x2004EE80 STACK= 760/1104 LOAD= 1.6%
OTG1          PRI= 60 sp=0x2004E588 STACK= 256/ 656 LOAD= 0.5%
monitor       PRI=183 sp=0x20020860 STACK= 480/ 848 LOAD= 0.9%
timer         PRI=181 sp=0x20021AE0 STACK=1480/1872 LOAD= 2.7%
rcout         PRI=181 sp=0x200211A0 STACK= 440/ 848 LOAD=18.6%
rcin          PRI=177 sp=0x20020C00 STACK= 416/1360 LOAD= 9.3%
io            PRI= 58 sp=0x2001FEC0 STACK=1496/2384 LOAD= 0.0%*
storage       PRI= 59 sp=0x20021540 STACK= 824/1360 LOAD= 0.4%
UART1         PRI= 60 sp=0x2004DE20 STACK= 256/ 656 LOAD= 0.2%
UART3         PRI= 60 sp=0x2004D728 STACK= 256/ 656 LOAD= 0.2%
UART6         PRI= 60 sp=0x2004D030 STACK= 296/ 656 LOAD= 0.7%
UART7         PRI= 60 sp=0x2004C938 STACK= 368/ 656 LOAD= 0.2%
I2C0          PRI=176 sp=0x2004AFB8 STACK= 888/1360 LOAD= 1.1%
OSD           PRI= 59 sp=0x20046D48 STACK= 920/1616 LOAD= 0.2%
UART2         PRI= 60 sp=0x20047A80 STACK= 240/ 656 LOAD= 0.2%
log_io        PRI= 59 sp=0x20047D78 STACK=1040/1656 LOAD= 0.5%
SPI4          PRI=181 sp=0x20039170 STACK= 584/1360 LOAD=27.2%
FTP           PRI= 58 sp=0x20034940 STACK=1432/2896 LOAD= 4.1%
```
after:
```
ThreadsV2
ISR           PRI=255 sp=0x20010000 STACK=1240/1536 LOAD= 3.7%
ArduCopter    PRI=182 sp=0x20010600 STACK=5352/7168 LOAD=43.5%
idle          PRI=  1 sp=0x20023770 STACK= 144/ 352 LOAD= 1.1%
UART_RX       PRI= 60 sp=0x2004EE80 STACK= 760/1104 LOAD= 0.9%
OTG1          PRI= 60 sp=0x2004E588 STACK= 256/ 656 LOAD= 0.2%
monitor       PRI=183 sp=0x20020860 STACK= 480/ 848 LOAD= 0.6%
timer         PRI=181 sp=0x20021AE0 STACK=1480/1872 LOAD= 1.7%
rcout         PRI=181 sp=0x200211A0 STACK= 424/ 848 LOAD=14.6%
rcin          PRI=177 sp=0x20020C00 STACK= 440/1360 LOAD= 6.9%*
io            PRI= 58 sp=0x2001FEC0 STACK=1496/2384 LOAD= 0.4%
storage       PRI= 59 sp=0x20021540 STACK= 864/1360 LOAD= 0.2%
UART1         PRI= 60 sp=0x2004DE20 STACK= 304/ 656 LOAD= 0.1%
UART3         PRI= 60 sp=0x2004D728 STACK= 256/ 656 LOAD= 0.1%
UART6         PRI= 60 sp=0x2004D030 STACK= 296/ 656 LOAD= 0.4%
UART7         PRI= 60 sp=0x2004C938 STACK= 368/ 656 LOAD= 0.1%
I2C0          PRI=176 sp=0x2004AFB8 STACK= 888/1360 LOAD= 0.7%
OSD           PRI= 59 sp=0x20046D48 STACK= 920/1616 LOAD= 0.5%
UART2         PRI= 60 sp=0x20047800 STACK= 240/ 656 LOAD= 0.2%
log_io        PRI= 59 sp=0x20047D78 STACK=1040/1656 LOAD= 0.9%
SPI4          PRI=181 sp=0x20037E48 STACK= 568/1360 LOAD=22.4%
FTP           PRI= 58 sp=0x20032030 STACK=1432/2896 LOAD= 0.3%
```

### CODE REVIEW

**Comment by IamPete1** [comment]
This is a relevantly big flash hit on F4. 
```
Binary           text             data         bss          total
---------------  ---------------  -----------  -----------  ---------------
ardurover        3472 (+0.4163%)  0 (0.0000%)  0 (0.0000%)  3472 (+0.3597%)
arducopter       3496 (+0.3736%)  0 (0.0000%)  0 (0.0000%)  3496 (+0.3277%)
antennatracker   3472 (+0.5290%)  0 (0.0000%)  0 (0.0000%)  3472 (+0.4409%)
ardusub          3472 (+0.4315%)  0 (0.0000%)  0 (0.0000%)  3472 (+0.3710%)
arducopter-heli  3496 (+0.3796%)  0 (0.0000%)  0 (0.0000%)  3496 (+0.3323%)
arduplane        3544 (+0.3646%)  0 (0.0000%)  0 (0.0000%)  3544 (+0.3213%)
blimp            3472 (+0.5141%)  0 (0.0000%)  0 (0.0000%)  3472 (+0.4305%)
```

**Comment by andyp1per** [comment]
@IamPete1 yes it does - but the performance hit since the summer is pretty bad IMO - we're clearly paying the price in general flight for all the added features, doesn't matter so much on H7 but on older MCUs people will be noticing the difference. I'll see whether I can play around with which files has the biggest impact at the lowest cost. The SPI cost is common across all MCUs I've looked at and will be down to the filtering that we run in the backend thread.

**Comment by andyp1per** [comment]
We could consider doing this only on copter

**Comment by khancyr** [comment]
It would be great to not have to much micro optimisation like this as it make debugging harder too

**Comment by andyp1per** [comment]
There is something deeper going on here, I am papering over the cracks by reducing the load of dshot and SPI but it appears that the main thread can get heavily overloaded in certain circumstances, but I can't work out what causes it.

**Comment by andyp1per** [comment]
After lots of experimentation and looking at callgrind I get this before/after in stabilize:
```
ISR           PRI=255 sp=0x20010000 STACK=1224/1536 LOAD= 3.7%
ArduCopter    PRI=182 sp=0x20010600 STACK=5336/7168 LOAD=42.6%
idle          PRI=  1 sp=0x200237A8 STACK= 144/ 352 LOAD= 3.2%
UART_RX       PRI= 60 sp=0x2004EE80 STACK= 760/1104 LOAD= 1.3%
OTG1          PRI= 60 sp=0x2004E588 STACK= 256/ 656 LOAD= 0.2%
monitor       PRI=183 sp=0x20020890 STACK= 480/ 848 LOAD= 0.6%
timer         PRI=181 sp=0x20021B10 STACK=1480/1872 LOAD= 1.7%
rcout         PRI=181 sp=0x200211D0 STACK= 432/ 848 LOAD=14.5%
rcin          PRI=177 sp=0x20020C30 STACK= 464/1360 LOAD= 6.5%*
io            PRI= 58 sp=0x2001FEF0 STACK=1496/2384 LOAD= 0.7%
storage       PRI= 59 sp=0x20021570 STACK= 864/1360 LOAD= 0.3%
UART1         PRI= 60 sp=0x2004DE20 STACK= 312/ 656 LOAD= 0.2%
UART3         PRI= 60 sp=0x2004D728 STACK= 248/ 656 LOAD= 0.2%
UART6         PRI= 60 sp=0x2004D030 STACK= 296/ 656 LOAD= 0.6%
UART7         PRI= 60 sp=0x2004C938 STACK= 368/ 656 LOAD= 0.2%
I2C0          PRI=176 sp=0x2004AEF0 STACK= 888/1360 LOAD= 0.8%
OSD           PRI= 59 sp=0x20046C80 STACK= 920/1616 LOAD= 0.6%*
UART2         PRI= 60 sp=0x20047738 STACK= 240/ 656 LOAD= 0.2%
log_io        PRI= 59 sp=0x20047CB0 STACK=1040/1656 LOAD= 0.9%
SPI4          PRI=181 sp=0x20038D78 STACK= 704/1360 LOAD=20.3%
FTP           PRI= 58 sp=0x20035818 STACK=1432/2896 LOAD= 0.4%
```
and
```
ISR           PRI=255 sp=0x20010000 STACK=1264/1536 LOAD= 3.6%
ArduCopter    PRI=182 sp=0x20010600 STACK=5360/7168 LOAD=42.9%
idle          PRI=  1 sp=0x200237A8 STACK= 144/ 352 LOAD= 1.4%
UART_RX       PRI= 60 sp=0x2004EE80 STACK= 760/1104 LOAD= 1.0%
OTG1          PRI= 60 sp=0x2004E588 STACK= 256/ 656 LOAD= 0.2%
monitor       PRI=183 sp=0x20020890 STACK= 480/ 848 LOAD= 0.6%
timer         PRI=181 sp=0x20021B10 STACK=1480/1872 LOAD= 1.7%
rcout         PRI=181 sp=0x200211D0 STACK= 440/ 848 LOAD=14.6%
rcin          PRI=177 sp=0x20020C30 STACK= 424/1360 LOAD= 6.5%*
io            PRI= 58 sp=0x2001FEF0 STACK=1496/2384 LOAD= 0.5%
storage       PRI= 59 sp=0x20021570 STACK= 848/1360 LOAD= 0.3%
UART1         PRI= 60 sp=0x2004DE20 STACK= 312/ 656 LOAD= 0.2%
UART3         PRI= 60 sp=0x2004D728 STACK= 240/ 656 LOAD= 0.1%
UART6         PRI= 60 sp=0x2004D030 STACK= 296/ 656 LOAD= 0.4%
UART7         PRI= 60 sp=0x2004C938 STACK= 368/ 656 LOAD= 0.1%
I2C0          PRI=176 sp=0x2004AEF8 STACK= 888/1360 LOAD= 0.8%
UART2         PRI= 60 sp=0x20046990 STACK= 240/ 656 LOAD= 0.2%
OSD           PRI= 59 sp=0x20047098 STACK= 920/1616 LOAD= 0.5%
log_io        PRI= 59 sp=0x200479D0 STACK=1040/1656 LOAD= 0.9%
SPI4          PRI=181 sp=0x20038DC8 STACK= 584/1360 LOAD=22.4%
FTP           PRI= 58 sp=0x200328E8 STACK=1432/2896 LOAD= 0.4%
```
So bar far the biggest bang for the buck is optimizing the SPI thread which means optimizing the backend and the filter code. Everything else is lost in the noise.

**Review by khancyr — COMMENTED** [review]
You should probably check for the DEBUG and SITL build there to be able to disable the optimisation on debug build. There should a example in EKF scope as example

**Comment by andyp1per** [comment]
This now down to <300 bytes which seems pretty good for the performance improvement. Interestingly it was the vector ops in the SPI loop that were the most costly.

File: libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#define INLINE_VECTOR_OPS
```
Comment (tridge) [inline libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:1]: AP_INLINE_VECTOR_OPS

Comment (andyp1per) [reply libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp]: Fixed

**Comment by andyp1per** [comment]
Turns out temp cal is another big hog. Inlining vector ops in this file saves another 0.5% or so. Also confirmed that this costs 288 bytes as per the size test.

**Comment by tridge** [comment]
> Turns out temp cal is another big hog. Inlining vector ops in this file saves another 0.5% or so

we should probably just do the temp correction calculation only once per 100 samples. Temperatures don't change very fast

**Comment by andyp1per** [comment]
> we should probably just do the temp correction calculation only once per 100 samples. Temperatures don't change very fast

Yes agreed! Makes a lot of sense

**Comment by andyp1per** [comment]
@tridge it turns out that inline ops gets all the benefit, doing it less makes an indistinguishable benefit. So doesn't seem like the complexity is worth it.

**Review by tridge — APPROVED** [review]

**Pull request merged**